olive/passes/onnx/extract_adapters.py

```diff
@@ -0,0 +1,336 @@
+from olive.passes.pass_config import PassConfigParam
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+logger = logging.getLogger(__name__)
+
+
+class ExtractAdapters(Pass):
+    """Extract adapter weights from model and save them as external weights file.
+
+    If make_inputs is False, model proto is invalid after this pass as the adapter weights point to non-existent
+    external files. Inference session must be created by first loading the adapter weights using
+    SessionOptions.add_external_initializers.
+
+    If make_inputs is True, the adapter weights are inputs to the model and must be provided during inference.
+    """
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "make_inputs": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description=(
+                    "Convert adapter weights to inputs. If false, the adapter weights will be set as initializers with"
+                    " external data."
+                ),
+            ),
+            "pack_inputs": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description=(
+                    "Pack adapter weights for the same module type into a single input tensor. Only used if make_inputs"
+                    " is True."
+                ),
+            ),
+        }
+
+    def _run_for_config(
+        self, model: ONNXModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModelHandler:
+        output_model_path = resolve_onnx_path(output_model_path, Path(model.model_path).name)
+
+        if "lora_modules" not in model.model_attributes:
+            raise ValueError("Model does not contain lora_modules attribute")
+
+        # create a dag from the model
+        dag = OnnxDAG.from_model_path(model.model_path)
+        # remove unnecessary identity nodes
+        self.remove_identity_nodes(dag)
+
+        # get lora modules
+        lora_modules = model.model_attributes["lora_modules"]
+        lora_name_patterns = self._get_lora_name_patterns(lora_modules)
+
+        # dictionary to store adapter weights
+        weights = {}
+        # keep track of float and quantized modules
+        float_modules = set()
+        quant_modules = set()
+
+        # nodes to remove at the end
+        nodes_to_remove = set()
+
+        for node_name in list(dag.nodes.keys()):
+            if dag.get_node_op_type(node_name) != "MatMul" or not any(
+                re.match(pattern, node_name) for pattern in lora_name_patterns
+            ):
+                # not a lora module
+                continue
+
+            # new name for the weight
+            new_weight_name = self._create_new_weight_name(node_name)
+
+            # original weight name
+            old_weight_name = dag.nodes[node_name].inputs[1]
```

Comment:
dag.nodes.items()?

Reply:
I missed updating this. it is supposed to be `dag.get_node_inputs(node_name)[1]`. 

Users of `OnnxDAG` are supposed to use the public methods and not deal with the graph directly. So that they are agnostic to how the graph is represented internally.

Reply:
updated.